packages/opentelemetry-exporter-prometheus/src/prometheus.ts

```diff
@@ -184,7 +184,9 @@ export class PrometheusExporter implements MetricExporter {
      */
     if (metric instanceof Counter) {
       metric.remove(
-        ...record.descriptor.labelKeys.map(k => record.labels[k].toString())
+        ...record.descriptor.labelKeys.map(k =>
```

Comment:
In what cases would `record.labels[k]` be missing?

Reply:
In my manual testing last night this change prevented an error being thrown when a user instantiates a counter with a label key

    const monotonicCounter = meter.createCounter('monotonic_counter', {
      monotonic: true,
      labelKeys: ['pid'],
      description: 'Example of a monotonic counter',
    });

but then fails to provide a value for that key.  

    monotonicCounter.add(1, {})
    monotonicCounter.add(1)    

@dyladan With my build issues sorted/understood I can (depending on how your conversation with @mayurkale22 goes) either add tests that make the use-case here clear or revert this.

Reply:
Ah I understand. Instead of falling back to an empty string, I think I would prefer a filter on the list.

```typescript
      metric.remove(
        ...record.descriptor.labelKeys
          .filter(k => record.labels[k] != null)
          .map(k => record.labels[k].toString())
      );
```

Reply:
This is clever solution but feel like it won't work as expected.

AFAIK prom-client will throw "**Invalid number of arguments**" exception if length of keys != length of values. See: https://github.com/siimon/prom-client/blob/master/lib/util.js#L34

One option is to fall back on "**value missing** or **unknown**" as a default value when the corresponding value is not defined in any given key(s), similar to the [description field](https://github.com/open-telemetry/opentelemetry-js/blob/8a29221c08a90fd6fe85970210b4c3c8a3867de7/packages/opentelemetry-exporter-prometheus/src/prometheus.ts#L200). WDYT?

Reply:
Option 2: Instead of [blindly attaching](https://github.com/open-telemetry/opentelemetry-js/blob/master/packages/opentelemetry-exporter-prometheus/src/prometheus.ts#L201) our `Labelkeys` to prom-client's `labelNames`, we can filter the keys based on valid value.

This is what python has implemented: https://github.com/open-telemetry/opentelemetry-python/blob/master/ext/opentelemetry-ext-prometheus/src/opentelemetry/ext/prometheus/__init__.py#L146-L150

Reply:
Thank you for your feedback @mayurkale22

> WDYT?

IT I've been wrangled into an bug fix that is _just_ outside the scope of my original PR -- but I can live with that :)

Re: option 2 -- when you say 

> Option 2: Instead of blindly attaching our Labelkeys to prom-client's labelNames, we can filter the keys based on valid value.

My interpretation of this is it's a suggestion that, when it comes time to create the Prometheus metric, we look at the keys and values of the passed in Open Telemetry metric, remove any keys with "non-valid" values (undefined, null, -- others?) and create the Prometheus metric with this new, smaller, set of keys. 

If that's the suggestion -- I don't think this is the right course of action.  Label keys seem like a fundamental part of a metric.  That is -- a metric with the keys [`foo`, `bar`, `baz`] seems like a _different_ metric than one with the keys [`foo`, `bar`].  Removing keys doesn't seems like the right course of action. 

Give than, Option 1 seems like the better approach.  

> One option is to fall back on "value missing or unknown" as a default value when the corresponding value is not defined in any given key(s),

You didn't indicate whether you thought this value should be added when creating the Prometheus metric, or when _recording_ the Open Telemetry metric.  My assumption is the former, as having this extra work happen _every_ time a metric is recorded would be excessive. 

Given all that: I intend to change how Prometheus metrics are instantiated.  If a value isn't present for a specific key, I will use a constant string of `value missing` for its value.  I will add tests to cover this case.  I'll check the spec for what constitutes a valid values, and make judgment call if there are ambiguities there.

Reply:
@dyladan would like you know your thoughts on this? 

> IT I've been wrangled into an bug fix that is just outside the scope of my original PR

Perhaps we can merge this PR and address prometheus exporter issue in the separate PR (btw, this won't make master branch unstable but we need to apply fix before next release). If we decide to merge, I would suggest to revert prometheus related to changes from this PR.

Reply:
Think I would rather see it as two separate PRs simply for the purpose of having better history later when we have to go spelunking through PRs to discover why decisions were made. Not a strong preference though. Agree the prom changes should be reverted if this is going to merge without a proper fix.